File: Dynamic Programming/distance_of_nearest_0.cpp
/*Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

The distance between two adjacent cells is 1.

 

Example 1:


Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]
Example 2:


Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]
 

Constraints:

m == mat.length
n == mat[i].length
1 <= m, n <= 10^4
1 <= m * n <= 10^4
mat[i][j] is either 0 or 1.
There is at least one 0 in mat.


explanation: 

stepwise explanation of the code: 

1. We start by initializing the dimensions of the input matrix mat as m (number of rows) and n (number of columns). This will be used later in the code.

2. We create a result matrix called result of size m x n to store the distance of the nearest 0 for each cell. Initially, we set all the values in result to a large value (in this case, INT_MAX) to represent an unreachable distance.

3. We create a queue called q to store the cell indices that need to be processed during the breadth-first search.

4. We iterate through the input matrix mat and enqueue the indices of all the cells with value 0 into the queue. Additionally, we mark the distance of these cells in the result matrix as 0. This initialization step ensures that the cells with value 0 are considered as starting points for the breadth-first search.

5. Now, we perform the breadth-first search using the queue q. While the queue is not empty, we process the cells in a breadth-first manner.

6. For each cell (row, col) popped from the queue, we check its neighboring cells in four directions: up, down, left, and right. We define the directions as a vector of pairs called directions, where each pair represents the change in row and column indices to move in a specific direction.

7. If the neighboring cell indices (newRow, newCol) are within the valid range of the matrix and the current distance in the result matrix for (newRow, newCol) is greater than the distance of the current cell plus 1, we update the distance in the result matrix and enqueue the neighboring cell (newRow, newCol) into the queue for further processing.

8. Once the breadth-first search is completed, the result matrix will contain the distance of the nearest 0 for each cell in the input matrix.

9. Finally, we return the result matrix.

*/

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution
{
public:
    vector<vector<int>> updateMatrix(vector<vector<int>> &mat)
    {
        int m = mat.size();
        int n = mat[0].size();

        // Create a result matrix and initialize with large values
        vector<vector<int>> result(m, vector<int>(n, INT_MAX));

        // Create a queue to store cell indices
        queue<pair<int, int>> q;

        // Initialize the queue with 0 cells and mark their distance as 0
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (mat[i][j] == 0)
                {
                    result[i][j] = 0;
                    q.push({i, j});
                }
            }
        }

        // Perform breadth-first search
        vector<pair<int, int>> directions{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        while (!q.empty())
        {
            int row = q.front().first;
            int col = q.front().second;
            q.pop();

            for (const auto &dir : directions)
            {
                int newRow = row + dir.first;
                int newCol = col + dir.second;

                if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n)
                {
                    if (result[newRow][newCol] > result[row][col] + 1)
                    {
                        result[newRow][newCol] = result[row][col] + 1;
                        q.push({newRow, newCol});
                    }
                }
            }
        }

        return result;
    }
};

int main()
{
    // Example usage
    vector<vector<int>> mat = {{0, 0, 0}, {0, 1, 0}, {0, 0, 0}};

    Solution s;
    vector<vector<int>> result = s.updateMatrix(mat);

    // Print the result
    for (const auto &row : result)
    {
        for (int val : row)
        {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: Dynamic Programming/distance_of_nearest_0.py
'''
Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.
The distance between two adjacent cells is 1.
 
Example 1:
Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]
Example 2:
Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]
 
Constraints:
m == mat.length
n == mat[i].length
1 <= m, n <= 104
1 <= m * n <= 104
mat[i][j] is either 0 or 1.
There is at least one 0 in mat.

Intuition
The goal is to get the nearest 0 for each 1. So how can we get the nearest 0 for each 1? Lets use Multisource BFS and Dynamic Programming.
Approach
To get the nearest 0 for each 1, we should treat every 0 as if they are on the same level, make a BFS on them and keep track of the distance as we go further. Since we are finding the nearest zero, we may need to use pre-calculated distances for the current 1 when all of its neighbours are all 1.
Lets see this with example!
Here is the given grid.
    0 0 0         
    0 1 0         
    1 1 1   
Then by taking all zeros and making a BFS on them, the nearest distance becomes 1 for all ones other than the back ticked 1 found on the third row.
    0 0 0         
    0 1 0         
    1 `1` 1  
The back ticked '1' is not adjacent to 0, so it should use the precalculated distance of the nearest 1 it get.
So, in here we are using precalculated values to get the nearest distance. This technique is called Dynamic programming.
Then we will get this grid by using the values we calculated for its adjacent 1
    0 0 0         
    0 1 0         
    1 2 1  
Complexity
Time complexity: O(N)
Space complexity: O(V + E)
Code
sample input and output:

1. Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
   Output: [[0,0,0],[0,1,0],[0,0,0]]

'''
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        
        # it a Multi-source BFS like 994. Rotting Oranges
        # treat every 0 as if they are on the same level and make BFS on them
        def isvalid(row, col):
            if row >= len(mat) or row < 0 or col >= len(mat[0]) or col < 0:
                return False
            return True
        
        # get every zero and add them to a queue
        q, visited = deque(), set()
        for row in range( len(mat)):
            for col in range( len(mat[0])):
                if mat[row][col] == 0:
                    q.append([row, col])
                    visited.add((row, col))
        
        level = 1 # the distance to the nearest zero starts from 1
        while q:
            size = len(q)
            for _ in range( size ):
                row, col =q.popleft()
                for r, c in [ [1, 0], [-1, 0], [0, 1], [0, -1]]:
                    newRow, newCol = row + r, col + c
                    if (newRow, newCol) not in visited and isvalid(newRow, newCol):
                        mat[newRow][newCol] = level
                        q.append([newRow, newCol])
                        visited.add( (newRow, newCol))
            level += 1
        
        return mat